### PR TITLE
Initial refactoring effort: extract high-level sections into modules

### PR DESCRIPTION
**Pull request by redsquirrel**
Based on [this comment](https://github.com/ethereum/browser-solidity/pull/33#discussion_r64923904) by @chriseth, I'm attempting to refactor app.js. This is an initial attempt, mainly to see if the approach makes sense. The approach is to extract methods and concepts into cohesive modules in order to make dependencies clearer, and to eventually pare down app.js to something that can more easily fit into someone's head.

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -81,134 +33,59 @@ var run = function() {
 			}
 			window.localStorage[key] = content;
 		}
-		SOL_CACHE_FILE = fileKey(Object.keys(files)[0]);
+		editor.setCacheFile(Object.keys(files)[0]);
```
Comment (chriseth) [inline src/app.js]: That function should be named differently, but can / should be done in a subsequent refactoring step.

Comment (redsquirrel) [inline src/app.js]: Agreed, there's a ton more to clean up, rename, and clarify. For now, I tried to keep the names as similar as possible.

File: src/app.js [hunk 2]
```diff
@@ -81,134 +33,58 @@ var run = function() {
...
-		if(loadingFromGist) return;
-		// Backwards-compatibility
-		while (window.localStorage[SOL_CACHE_UNTITLED + untitledCount])
-			untitledCount = (untitledCount - 0) + 1;
-		SOL_CACHE_FILE = SOL_CACHE_UNTITLED + untitledCount;
-		window.localStorage[SOL_CACHE_FILE] = window.localStorage['sol-cache'] || BALLOT_EXAMPLE;
-		window.localStorage.removeItem('sol-cache');
-	}
+	var loadingFromGist = gistHandler.handleLoad(function(gistId) {
+		$.ajax({
```
Comment (chriseth) [inline src/app.js]: You mixed tabs and spaces here, but we should probably just run a formatter over this...

File: src/app.js [hunk 3]
```diff
@@ -81,134 +33,58 @@ var run = function() {
...
-	SOL_CACHE_FILE = getFiles()[0];
 
+	// ----------------- storage --------------------
 
-	var files = getFiles();
-	for (var x in files) {
-		sessions[files[x]] = newEditorSession(files[x]);
-	}
+	var storageHandler = new StorageHandler(updateFiles);
+	window.syncStorage = storageHandler.sync;
```
Comment (chriseth) [inline src/app.js]: I know it was like that before, but why is that needed?

File: src/app.js [hunk 4]
```diff
@@ -373,75 +194,53 @@ var run = function() {
 	$filesEl.on('click', '.file .remove', function(ev) {
 		ev.preventDefault();
 		var name = $(this).parent().find('.name').text();
-		var index = getFiles().indexOf( fileKey(name) );
 
 		if (confirm("Are you sure you want to remove: " + name + " from local storage?")) {
-			window.localStorage.removeItem( fileKey( name ) );
-			SOL_CACHE_FILE = getFiles()[ Math.max(0, index - 1)];
+			window.localStorage.removeItem( utils.fileKey( name ) );
+			editor.setNextFile(utils.fileKey(name));
```
Comment (chriseth) [inline src/app.js]: Isn't this rather `setCacheFileByName`?

File: src/app.js [hunk 5]
```diff
@@ -373,75 +194,53 @@ var run = function() {
...
 		ev.preventDefault();
-		SOL_CACHE_FILE = fileKey( $(this).find('.name').text() );
+		editor.setCacheFile($(this).find('.name').text());
 		updateFiles();
 		return false;
 	}
 
-	function fileTabFromKey(key) {
-		var name = fileNameFromKey(key);
+	function fileTabFromKey() {
```
Comment (chriseth) [inline src/app.js]: If you remove the parameter I think you should also rename it: `activeFileTab`?

File: src/app/compiler.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+  var cachedRemoteFiles = {};
+  var worker = null;
+
+  var compileTimeout = null;
+
+  function onChange() {
+    var input = editor.getValue();
+    if (input === "") {
+      window.localStorage.setItem(editor.getRawCacheFile(), '');
```
Comment (chriseth) [inline src/app/compiler.js]: That should probably move to the editor.

File: src/universal-dapp.js [hunk 7]
```diff
@@ -192,7 +192,6 @@ UniversalDApp.prototype.getInstanceInterface = function (contract, address, $tar
                 return;
 
             $event = $('<div class="event" />');
-
```
Comment (chriseth) [inline src/universal-dapp.js]: Not sure if this commit was correctly applied...

File: src/app.js (hunk 2 above)
Comment (redsquirrel) [inline src/app.js]: Do you want me to handle that now?

File: src/app.js (hunk 3 above)
Comment (redsquirrel) [inline src/app.js]: I have no clue what `window.syncStorage` does. I'm just trying to preserve behavior since this is a refactoring.

File: src/app.js (hunk 4 above)
Comment (redsquirrel) [inline src/app.js]: No, because it's using the name to find the *next* file. (Since the named file has been removed.) There are a flurry of subsequent minor refactorings to do once we get this major one out of the way.

File: src/universal-dapp.js (hunk 7 above)
Comment (redsquirrel) [inline src/universal-dapp.js]: Every line that @axic changed in app.js was extracted to the new modules. In the process, I made the same formatting updates. (Except for jsonInterface vs interface, which is about to be committed.)

**Comment by redsquirrel** [comment]
⚠️  DO NOT MERGE YET. ⚠️ 

I just found a bug. 🐜

**Comment by chriseth** [comment]
Looks good to me up to this point.

**Comment by redsquirrel** [comment]
OK, false alarm on the bug. Ready for merge.
